### ISSUE TITLE
freebsd build failure report

### ISSUE THREAD

**Issue by sezero** [Mindwerks/wildmidi#171]
See:
https://www.mail-archive.com/freebsd-pkg-fallout@freebsd.org/msg500763.html
https://www.mail-archive.com/freebsd-pkg-fallout@freebsd.org/msg501575.html

```
/wrkdirs/usr/ports/audio/wildmidi/work/wildmidi-wildmidi-0.4.0/src/wm_tty.c:54:22:
 error: use of undeclared identifier 'ONLCR'
    _tty.c_oflag &= ~ONLCR;
                     ^
1 error generated.
```

ONLCR is defined here:
https://svnweb.freebsd.org/base/head/sys/sys/_termios.h?view=log
... which seems to have had a recent change.

The infernal feature selection macros are here:
https://svnweb.freebsd.org/base/head/sys/sys/cdefs.h?view=log

**Comment by liangtai** [comment]
Hi,
I am sorry having no idea to work for these trouble reports,
though I am the maintainer of this port.

**Comment by chrisisonwildcode** [comment]
I didn't write wm_tty.c so I'll have to look over the code. I think Bret and I have the ability to build in a FreeBSD environment so testing in that sense is not an issue.

If anyone has any idea about these reports but unable to test then please give us a heads up and we can test for you.


Chris

**Comment by psi29a** [comment]
Not this one again... this was fixed awhile ago, seems to be a moving target. I managed to test on a kfreebsd/debian setup but that likely is a lot different than straight up freeBSD.

**Comment by sezero** [comment]
> Not this one again...

Yeah :)

**Comment by psi29a** [comment]
Install freebsd now...

**Comment by psi29a** [comment]
Ugh... guys, I can't duplicate this on FreeBSD 11.0

Am I doing something wrong? (or right?)

```
$ git clone https://github.com/Mindwerks/wildmidi.git
Cloning into 'wildmidi'...
remote: Counting objects: 5930, done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 5930 (delta 0), reused 1 (delta 0), pack-reused 5925
Receiving objects: 100% (5930/5930), 2.39 MiB | 675.00 KiB/s, done.
Resolving deltas: 100% (4239/4239), done.
$ mkdir build
$ cd build/
$ cmake ../wildmidi/
-- The C compiler identification is Clang 3.8.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Setting build type to 'Debug' as none was specified.
-- Build Type: Debug
-- Performing Test HAVE_VISIBILITY_DEFAULT
-- Performing Test HAVE_VISIBILITY_DEFAULT - Success
-- Performing Test HAVE_VISIBILITY_HIDDEN
-- Performing Test HAVE_VISIBILITY_HIDDEN - Success
-- Performing Test HAVE___BUILTIN_EXPECT
-- Performing Test HAVE___BUILTIN_EXPECT - Success
-- Performing Test HAVE_C_INLINE
-- Performing Test HAVE_C_INLINE - Success
-- Performing Test HAVE_C___INLINE__
-- Performing Test HAVE_C___INLINE__ - Success
-- Performing Test HAVE_C___INLINE
-- Performing Test HAVE_C___INLINE - Success
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- Could NOT find ALSA (missing:  ALSA_LIBRARY ALSA_INCLUDE_DIR) 
-- Could NOT find OpenAL (missing:  OPENAL_LIBRARY OPENAL_INCLUDE_DIR) 
-- Looking for OSS...
-- Looking for include file sys/soundcard.h
-- Looking for include file sys/soundcard.h - found
-- Looking for include file machine/soundcard.h
-- Looking for include file machine/soundcard.h - not found
-- Looking for include file soundcard.h
-- Looking for include file soundcard.h - not found
-- Performing Test OSS_FOUND
-- Performing Test OSS_FOUND - Success
-- Found OSS.
-- Configuring done
-- Generating done
-- Build files have been written to: /usr/home/psi29a/build
$ make
Scanning dependencies of target libwildmidi_dynamic
[  5%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/wm_error.c.o
[ 10%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/file_io.c.o
[ 15%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/lock.c.o
[ 20%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/wildmidi_lib.c.o
[ 25%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/reverb.c.o
[ 30%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/gus_pat.c.o
[ 35%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/internal_midi.c.o
[ 40%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/patches.c.o
[ 45%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/f_xmidi.c.o
[ 50%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/f_mus.c.o
[ 55%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/f_hmp.c.o
[ 60%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/f_midi.c.o
[ 65%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/f_hmi.c.o
[ 70%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/sample.c.o
[ 75%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/mus2mid.c.o
[ 80%] Building C object src/CMakeFiles/libwildmidi_dynamic.dir/xmi2mid.c.o
[ 85%] Linking C shared library ../libWildMidi.so
[ 85%] Built target libwildmidi_dynamic
Scanning dependencies of target wildmidi
[ 90%] Building C object src/CMakeFiles/wildmidi.dir/wm_tty.c.o
[ 95%] Building C object src/CMakeFiles/wildmidi.dir/wildmidi.c.o
[100%] Linking C executable ../wildmidi
[100%] Built target wildmidi
$ ./wildmidi

WildMidi 0.4.2 Open Source Midi Sequencer
Copyright (C) WildMIDI Developers 2001-2016

WildMidi comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it under
the terms and conditions of the GNU General Public License version 3.
For more information see COPYING

Report bugs to https://github.com/Mindwerks/wildmidi/issues
WildMIDI homepage is at http://www.mindwerks.net/projects/wildmidi/

ERROR: No midi file given
Usage: wildmidi [options] filename.mid
```

**Comment by sezero** [comment]
> Ugh... guys, I can't duplicate this on FreeBSD 11.0

Seems like the error reports are for head-amd64-default and head-i386-default,
development version of freebsd?

**Comment by psi29a** [comment]
OK... how do I set that up? ;)

**Comment by sezero** [comment]
> OK... how do I set that up? ;)

Don't look at me :)

**Comment by psi29a** [comment]
I've not had any success with the latest 12-CURRENT build, both boot-only and full disk and I can't seem to get it to install git/cmake via ports or pkg. It all times out. I had no problem with 11.0 however.

Can anyone test?

**Comment by liangtai** [comment]
This also happens on FreeBSD 11.1-PRERELEASE, the latest state of
11-STABLE, that introduced
the important change on src/sys/sys/_termios.h replacing POSIX_SOURCE with
__BSD_VISIBLE.

For a quick fix, just insert a line
  CFLAGS="-D__BSD_VISIBLE"
somewhere under the line
  .include <bsd.port.options.mk>
in audio/wildmidi/Makefile of your ports tree.

**Comment by psi29a** [comment]
Great, I'll try that next. Thanks. :)

**Comment by sezero** [comment]
I wonder what is preventing __BSD_VISIBLE from being defined in the first place.

**Comment by psi29a** [comment]
anyone know of a free CI environment we can use that is for freebsd? that would help find regressions faster. :)

**Comment by psi29a** [comment]
Normally we shouldn't be playing with double underscore defines, they are set internally by a combination of other options. However all the search researches indicate that I should use __BSD_VISIBLE instead of _XOPEN_SOURCE or _C_POSIXVERSION which is contradictory.

This compiles on freebsd-11.1-RC2, can someone else validate?

**Comment by sezero** [comment]
Backported the fix to wildmidi-0.3: commit cec16eff
